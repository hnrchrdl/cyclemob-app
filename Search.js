import React from 'react';
import {
  StyleSheet,
  FlatList,
  KeyboardAvoidingView,
  View,
  Text,
  TextInput,
  TouchableOpacity
} from 'react-native';
import {
  getPlacesAutocomplete,
  getPlaceDetails,
  createSessionToken
} from './lib';
import { debounce } from 'lodash';

class SearchResultItem extends React.Component {
  onPress = () => {
    const { item } = this.props;
    this.props.onItemSelect(item);
  };
  render() {
    const { item } = this.props;
    const { main_text } = item.structured_formatting;
    const { description } = item;
    return (
      <TouchableOpacity style={styles.resultItem} onPress={this.onPress}>
        <Text style={styles.resultItemName}>{main_text}</Text>
        <Text style={styles.resultItemVicinity}>{description}</Text>
      </TouchableOpacity>
    );
  }
}

class Search extends React.PureComponent {
  constructor() {
    super();
    this.state = {
      result: [],
      sessionToken: createSessionToken()
    };
  }
  onSearchTextChange = input => {
    if (input.length === 0) {
      this.setState({
        sessionToken: createSessionToken()
      });
    } else {
      getPlacesAutocomplete(input, this.state.sessionToken).then(data => {
        this.setState({ result: data });
      });
    }
  };

  onSearchResultSelect = item => {
    const { place_id } = item;
    getPlaceDetails(place_id, this.state.sessionToken).then(result => {
      this.props.onItemSelect(result);
    });
  };
  render() {
    return (
      <View style={styles.container}>
        <TextInput
          autoFocus
          style={styles.textInput}
          placeholder="City / POI"
          onChangeText={debounce(this.onSearchTextChange, 800)}
        />
        <View style={styles.resultContainer}>
          <FlatList
            data={this.state.result}
            renderItem={({ item }) => (
              <SearchResultItem
                key={item.place_id}
                item={item}
                onItemSelect={this.onSearchResultSelect}
              />
            )}
            keyboardShouldPersistTaps="handled"
          />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    padding: 20,
    flex: 1,
    borderTopColor: '#eee',
    borderTopWidth: 1,
    backgroundColor: '#fff'
  },
  resultContainer: {
    borderTopColor: '#eee',
    borderTopWidth: 1,
    marginTop: 15,
    paddingBottom: 15
  },
  textInput: {
    height: 40
  },
  resultItem: {
    marginTop: 10,
    paddingTop: 10
  },
  resultItemName: {},
  resultItemVicinity: {
    color: '#ccc'
  }
});

export default Search;
