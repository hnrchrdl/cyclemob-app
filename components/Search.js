import React from 'react';
import {
  StyleSheet,
  FlatList,
  View,
  Text,
  TextInput,
  TouchableOpacity
} from 'react-native';
import Button from './Button';
import PropTypes from 'prop-types';
import { getPlacesAutocomplete, getPlaceDetails } from '../lib/gmaps-api';
import { createUUID } from '../lib/helper';
import { debounce } from 'lodash';

class Search extends React.PureComponent {
  constructor() {
    super();
    this.state = {
      result: [],
      sessionToken: createUUID()
    };
  }
  static propTypes = {
    position: PropTypes.shape({
      coords: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number,
        speed: PropTypes.number,
        accuracy: PropTypes.number,
        altitude: PropTypes.number,
        heading: PropTypes.number
      }),
      mocked: PropTypes.bool,
      timestamp: PropTypes.number
    }).isRequired,
    onItemSelect: PropTypes.func.isRequired,
    onClose: PropTypes.func.isRequired
  };
  static ResultItem = ResultItem;

  onSearchTextChange = input => {
    if (input.length === 0) {
      this.setState({
        sessionToken: createUUID()
      });
    } else {
      const location = this.props.position.coords;
      getPlacesAutocomplete(input, location, this.state.sessionToken).then(
        data => {
          this.setState({ result: data });
        }
      );
    }
  };

  onSearchResultSelect = item => {
    const { place_id } = item;
    getPlaceDetails(place_id, this.state.sessionToken).then(result => {
      this.props.onItemSelect(result);
    });
  };
  render() {
    return (
      <View style={styles.container}>
        <View style={styles.inputContainer}>
          <View style={{ flex: 1 }}>
            <TextInput
              autoFocus
              style={styles.textInput}
              placeholder="City / POI"
              onChangeText={debounce(this.onSearchTextChange, 800)}
            />
          </View>
          <View style={{ flex: 0 }}>
            <Button iconName="close" onPress={this.props.onClose} />
          </View>
        </View>
        {this.state.result.length > 0 && (
          <View style={styles.resultContainer}>
            <FlatList
              data={this.state.result}
              renderItem={({ item }) => (
                <Search.ResultItem
                  key={item.place_id}
                  item={item}
                  onItemSelect={this.onSearchResultSelect}
                />
              )}
              keyboardShouldPersistTaps="handled"
            />
          </View>
        )}
      </View>
    );
  }
}

function ResultItem({ item, onItemSelect }) {
  return (
    <TouchableOpacity
      style={styles.resultItem}
      onPress={() => onItemSelect(item)}
    >
      <Text style={styles.resultItemName}>{item.structured_formatting}</Text>
      <Text style={styles.resultItemVicinity}>{item.description}</Text>
    </TouchableOpacity>
  );
}

ResultItem.propTypes = {
  item: PropTypes.shape({
    structured_formatting: PropTypes.string,
    description: PropTypes.string
  }).isRequired,
  onItemSelect: PropTypes.func.isRequired
};

const styles = StyleSheet.create({
  container: {
    padding: 10,
    backgroundColor: '#fff',
    opacity: 0.95
  },
  inputContainer: {
    flexDirection: 'row'
  },
  textInput: {
    paddingTop: 10,
    paddingBottom: 10,
    marginRight: 10
  },
  resultContainer: {
    borderTopColor: '#eee',
    borderTopWidth: 1,
    margin: 10
  },
  resultItem: {
    marginTop: 10,
    paddingTop: 10
  },
  resultItemName: {},
  resultItemVicinity: {
    color: '#ccc'
  }
});

export default Search;
