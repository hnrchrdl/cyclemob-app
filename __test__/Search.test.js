import React from 'react';
import Search from '../components/Search';

import { shallow } from 'enzyme';

jest.mock('TextInput', () => {
  const actualComponent = require.requireActual('TextInput');
  const React = require('React');
  class TextInput extends React.Component {
    render() {
      delete this.props.autoFocus;
      return React.createElement('TextInput', this.props, this.props.children);
    }
  }
  TextInput.propTypes = actualComponent.propTypes;
  return TextInput;
});

const props = {
  onItemSelect: jest.fn(),
  onClose: jest.fn(),
  position: { coords: { latitude: 50, longitude: 20 } }
};

it('renders without crashing', () => {
  const tree = shallow(<Search {...props} />);
  expect(tree).toBeTruthy();
});

it('matches snapshot', () => {
  const tree = shallow(<Search {...props} />);
  expect(tree).toMatchSnapshot();
});

it('should render a text input and react to user input', done => {
  const mockOnSearchTextChange = jest.fn();
  Search.prototype.onSearchTextChange = mockOnSearchTextChange;

  const tree = shallow(<Search {...props} />);
  const textInput = tree.find('TextInput');
  expect(textInput).toHaveLength(1);
  textInput.simulate('changeText', { target: { value: 'test' } });
  // debounced
  expect(mockOnSearchTextChange).not.toHaveBeenCalled();
  setTimeout(() => {
    expect(mockOnSearchTextChange).toHaveBeenCalledWith('test');
    done();
  }, 1500);
});

// Result Item

const resultItemProps = {
  item: {
    structured_formatting: 'some title',
    description: 'some description'
  },
  onItemSelect: jest.fn()
};

it('renders result item without crashing', () => {
  const tree = shallow(<Search.ResultItem {...resultItemProps} />);
  expect(tree).toBeTruthy();
});

it('matches snapshot for result item', () => {
  const tree = shallow(<Search.ResultItem {...resultItemProps} />);
  expect(tree).toBeTruthy();
});

it('should render result items correctly', () => {});
