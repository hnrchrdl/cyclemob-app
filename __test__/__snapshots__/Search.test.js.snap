// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    onClose={[MockFunction]}
    onItemSelect={[MockFunction]}
    position={
      Object {
        "coords": Object {
          "latitude": 50,
          "longitude": 20,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Component
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Component
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <TextInput
              autoFocus={true}
              onChangeText={[Function]}
              placeholder="City / POI"
              style={
                Object {
                  "marginRight": 10,
                  "paddingBottom": 10,
                  "paddingTop": 10,
                }
              }
            />
          </Component>
          <Component
            style={
              Object {
                "flex": 0,
              }
            }
          >
            <Button
              iconName="close"
              onPress={[MockFunction]}
            />
          </Component>
        </Component>,
        false,
      ],
      "style": Object {
        "backgroundColor": "#fff",
        "opacity": 0.95,
        "padding": 10,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Component
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <TextInput
                autoFocus={true}
                onChangeText={[Function]}
                placeholder="City / POI"
                style={
                  Object {
                    "marginRight": 10,
                    "paddingBottom": 10,
                    "paddingTop": 10,
                  }
                }
              />
            </Component>,
            <Component
              style={
                Object {
                  "flex": 0,
                }
              }
            >
              <Button
                iconName="close"
                onPress={[MockFunction]}
              />
            </Component>,
          ],
          "style": Object {
            "flexDirection": "row",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <TextInput
                autoFocus={true}
                onChangeText={[Function]}
                placeholder="City / POI"
                style={
                  Object {
                    "marginRight": 10,
                    "paddingBottom": 10,
                    "paddingTop": 10,
                  }
                }
              />,
              "style": Object {
                "flex": 1,
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "autoFocus": true,
                "onChangeText": [Function],
                "placeholder": "City / POI",
                "style": Object {
                  "marginRight": 10,
                  "paddingBottom": 10,
                  "paddingTop": 10,
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Button
                iconName="close"
                onPress={[MockFunction]}
              />,
              "style": Object {
                "flex": 0,
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "iconName": "close",
                "onPress": [MockFunction],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Component
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <Component
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <TextInput
                autoFocus={true}
                onChangeText={[Function]}
                placeholder="City / POI"
                style={
                  Object {
                    "marginRight": 10,
                    "paddingBottom": 10,
                    "paddingTop": 10,
                  }
                }
              />
            </Component>
            <Component
              style={
                Object {
                  "flex": 0,
                }
              }
            >
              <Button
                iconName="close"
                onPress={[MockFunction]}
              />
            </Component>
          </Component>,
          false,
        ],
        "style": Object {
          "backgroundColor": "#fff",
          "opacity": 0.95,
          "padding": 10,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Component
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <TextInput
                  autoFocus={true}
                  onChangeText={[Function]}
                  placeholder="City / POI"
                  style={
                    Object {
                      "marginRight": 10,
                      "paddingBottom": 10,
                      "paddingTop": 10,
                    }
                  }
                />
              </Component>,
              <Component
                style={
                  Object {
                    "flex": 0,
                  }
                }
              >
                <Button
                  iconName="close"
                  onPress={[MockFunction]}
                />
              </Component>,
            ],
            "style": Object {
              "flexDirection": "row",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <TextInput
                  autoFocus={true}
                  onChangeText={[Function]}
                  placeholder="City / POI"
                  style={
                    Object {
                      "marginRight": 10,
                      "paddingBottom": 10,
                      "paddingTop": 10,
                    }
                  }
                />,
                "style": Object {
                  "flex": 1,
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "autoFocus": true,
                  "onChangeText": [Function],
                  "placeholder": "City / POI",
                  "style": Object {
                    "marginRight": 10,
                    "paddingBottom": 10,
                    "paddingTop": 10,
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Button
                  iconName="close"
                  onPress={[MockFunction]}
                />,
                "style": Object {
                  "flex": 0,
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "iconName": "close",
                  "onPress": [MockFunction],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
